src/write/buf_write.rs

```diff
@@ -0,0 +1,29 @@
+use std::{
+    io,
+    pin::Pin,
+    task::{Context, Poll},
+};
+
+pub(crate) trait AsyncBufWrite {
+    /// Attempt to return an internal buffer to write to, flushing data out to the inner reader if
+    /// it is full.
+    ///
+    /// On success, returns `Poll::Ready(Ok(buf))`.
+    ///
+    /// If the buffer is full and cannot be flushed, the method returns `Poll::Pending` and
+    /// arranges for the current task context (`cx`) to receive a notification when the object
+    /// becomes readable or is closed.
+    fn poll_flush_buf(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<io::Result<&mut [u8]>>;
+
+    /// Tells this buffer that `amt` bytes have been written to its buffer, so they should be
+    /// written out to the underlying IO when possible.
```

Comment:
```suggestion
    /// written out to the underlying writer when flushed.
```

Reply:
This could be a direct `AsyncBufWrite` implementation on some IO device, it might not be a wrapper around another writer (I shouldn't have mentioned an "inner reader" earlier either).